### PR TITLE
[SPARK-36445][SQL] ANSI type coercion rule for date time operations

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Implement a new rule for the date-time operations in the ANSI type coercion system:
1. Date will be converted to Timestamp when it is in the subtraction with Timestmap.
2. Promote string literals in date_add/date_sub/time_add

### Why are the changes needed?

Currently the type coercion rule `DateTimeOperations` doesn't match the design of the ANSI type coercion system:
1. For date_add/date_sub, if the input is timestamp type, Spark should not convert it into date type since date type is narrower than the timestamp type.
2. For date_add/date_sub/time_add, string value can be implicit cast to date/timestamp only when it is literal.

Thus, we need to have a new rule for the date-time operations in the ANSI type coercion system.

### Does this PR introduce _any_ user-facing change?

No, the ANSI type coercion rules are not releaesd.

### How was this patch tested?

New UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 1]
```diff
@@ -289,6 +289,13 @@ object AnsiTypeCoercion extends TypeCoercionBase {
           case other => other
         }
         p.makeCopy(Array(a, newList))
+
+      case d @ DateAdd(left @ StringType(), _) if left.foldable =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:293]: shall we put these 3 in `DateTimeOperations`? The doc of this rule says `Promotes string literals that appear in arithmetic and comparison expressions.`.

File: sql/core/src/test/resources/sql-tests/inputs/date.sql [hunk 2]
```diff
@@ -91,6 +91,11 @@ create temp view v as select '1' str;
 select date_add('2011-11-11', str) from v;
 select date_sub('2011-11-11', str) from v;
 
+-- non-literal string column add/sub with integer
+create or replace temp view v2 as select '2011-11-11' str;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/date.sql:95]: nit: let's remove `or replace`. It's always error-prone if we replace a temp view in the test, as we may change the rest of the test cases.

File: sql/core/src/test/resources/sql-tests/inputs/timestamp.sql [hunk 3]
```diff
@@ -102,6 +102,8 @@ select timestamp'2011-11-11 11:11:11' - interval '2' second;
 select '2011-11-11 11:11:11' - interval '2' second;
 select '1' - interval '2' second;
 select 1 - interval '2' second;
+create or replace temp view v as select '2011-11-11' str;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:105]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/timestamp.sql [hunk 4]
```diff
@@ -102,6 +102,8 @@ select timestamp'2011-11-11 11:11:11' - interval '2' second;
...
+select str - interval '2' second from v;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:106]: -- non-literal string column subtract interval

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:106]: -- non-literal string column subtract interval

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:106]: how about string - date/timestamp?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:293]: Well, DateAdd is a sort of arithmetic expression

File: sql/core/src/test/resources/sql-tests/inputs/timestamp.sql [hunk 5]
```diff
@@ -102,6 +102,13 @@ select timestamp'2011-11-11 11:11:11' - interval '2' second;
 select '2011-11-11 11:11:11' - interval '2' second;
 select '1' - interval '2' second;
 select 1 - interval '2' second;
+select '2011-11-11 11:11:11' - date'2011-11-11';
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:105]: > how about string - date/timestamp?

I have added a new test case. But it seems that `string - date` will be `cast(string as date) - date`.  Should we move it to date.sql or just leave it here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:293]: I prefer to put all the string literal promotion here. Let me update the comment.

File: sql/core/src/test/resources/sql-tests/inputs/timestamp.sql (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:105]: let's put it in `date.sql`, as we don't need to test this query 4 times with different timestamp settings.

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out [hunk 6]
```diff
@@ -590,6 +590,58 @@ org.apache.spark.sql.AnalysisException
...
+0 00:00:00.000000000
+
+
+-- !query
+select '2011-11-11 11:11:11' - timestamp'2011-11-11 11:11:10'
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.AnalysisException
+cannot resolve '('2011-11-11 11:11:11' - TIMESTAMP '2011-11-11 11:11:10')' due to data type mismatch: argument 1 requires (timestamp or timestamp without time zone) type, however, ''2011-11-11 11:11:11'' is of string type.; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:607]: why does this fail under ansi mode?

File: sql/core/src/test/resources/sql-tests/results/datetime-legacy.sql.out [hunk 7]
```diff
@@ -1107,6 +1131,57 @@ org.apache.spark.sql.AnalysisException
...
+cannot resolve '('2011-11-11 11:11:11' - TIMESTAMP '2011-11-11 11:11:10')' due to data type mismatch: argument 1 requires (timestamp or timestamp without time zone) type, however, ''2011-11-11 11:11:11'' is of string type.; line 1 pos 7
+
+
+-- !query
+create temp view v as select '2011-11-11 11:11:11' str
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.catalyst.analysis.TempTableAlreadyExistsException
+Temporary view 'v' already exists
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/datetime-legacy.sql.out:1157]: seems we need to change the temp view name

File: sql/core/src/test/resources/sql-tests/results/timestamp.sql.out [hunk 8]
```diff
@@ -571,6 +571,56 @@ org.apache.spark.sql.AnalysisException
...
+0 00:00:00.000000000
+
+
+-- !query
+select '2011-11-11 11:11:11' - timestamp'2011-11-11 11:11:10'
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.AnalysisException
+cannot resolve '('2011-11-11 11:11:11' - TIMESTAMP '2011-11-11 11:11:10')' due to data type mismatch: argument 1 requires (timestamp or timestamp without time zone) type, however, ''2011-11-11 11:11:11'' is of string type.; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/timestamp.sql.out:588]: this fails in non-ansi mode as well... weird

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out (hunk 6 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:607]: We don't have a rule for this one

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:599]: shall we fix it under ansi mode? this is inconsistent: `string_literal - date` works but `string_literal - timestamp` does not.

**Comment by gengliangwang** [comment]
jenkins, retest this please

File: sql/core/src/test/resources/sql-tests/inputs/date.sql [hunk 9]
```diff
@@ -104,6 +110,7 @@ select date '2001-09-28' - null;
 select '2011-11-11' - interval '2' day;
 select null - date '2019-10-06';
 select date '2001-10-01' - date '2001-09-28';
+select '2011-11-11 11:11:11' - date'2011-11-11';
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/date.sql:114]: nit: we can put `select str - date'2011-11-11' from v2;` here.

File: sql/core/src/test/resources/sql-tests/results/date.sql.out [hunk 10]
```diff
@@ -331,12 +331,20 @@ struct<date_add(DATE '2011-11-11', 1):date>
...
-select date_add(timestamp_ltz'2011-11-11 12:12:12', 1), date_add(timestamp_ntz'2011-11-11 12:12:12', 1)
+select date_add(timestamp_ltz'2011-11-11 12:12:12', 1)
+-- !query schema
+struct<date_add(TIMESTAMP '2011-11-11 12:12:12', 1):date>
+-- !query output
+2011-11-12
+
+
+-- !query
+select date_add(timestamp_ntz'2011-11-11 12:12:12', 1)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/date.sql.out:342]: if `timestamp_ntz` becomes the default in the future, it's better to support this query as well to not break anything.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/results/date.sql.out (hunk 10 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/date.sql.out:342]: The LTZ one will fail in ANSI mode. We tend to make the NTZ behavior consistent in default/ANSI mode, right?

File: sql/core/src/test/resources/sql-tests/inputs/date.sql (hunk 9 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/inputs/date.sql:115]: Done

File: sql/core/src/test/resources/sql-tests/results/date.sql.out (hunk 10 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/date.sql.out:342]: It's a different topic. Breaking change means it's very hard to enable NTZ by default in the future.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/date.sql.out:342]: I think we tend to make LTZ/NTZ behavior consistent within default mode, or within ANSI mode, but not between default and ANSI mode.

**Comment by gengliangwang** [comment]
Merging to master/3.2

File: sql/core/src/test/resources/sql-tests/results/date.sql.out (hunk 10 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/date.sql.out:342]: I will have a follow-up for NTZ behavior.

**Pull request closed without merging**